--- a/sys/sys/elf32.h
+++ b/sys/sys/elf32.h
@@ -35,6 +35,7 @@
 #include <elf.h>
 #endif
 
+#if 0
 /*
  * ELF definitions common to all 32-bit architectures.
  */
@@ -150,6 +151,8 @@
  */
 typedef Elf_Note Elf32_Nhdr;
 
+#endif
+
 /*
  *	Move entry
  */
@@ -161,6 +164,7 @@
 	Elf32_Half	m_stride;	/* stride info */
 } Elf32_Move;
 
+#if 0
 /*
  *	The macros compose and decompose values for Move.r_info
  *
--- a/sys/sys/elf64.h
+++ b/sys/sys/elf64.h
@@ -52,6 +52,7 @@
 typedef uint64_t	Elf64_Xword;
 #endif
 
+#if 0
 /*
  * Types of dynamic symbol hash table bucket and chain elements.
  *
@@ -166,6 +167,8 @@
  */
 typedef Elf_Note Elf64_Nhdr;
 
+#endif
+
 /*
  *	Move entry
  */
@@ -183,6 +186,7 @@
 
 #endif
 
+#if 0
 /*
  *	Hardware/Software capabilities entry
  */
