--- a/sys/sys/limits.h
+++ b/sys/sys/limits.h
@@ -78,14 +78,14 @@
 #endif
 #endif /* 0 */
 
-#if __POSIX_VISIBLE >= 200112 || __XSI_VISIBLE
+#if 1
 #define	SIZE_T_MAX	__SIZE_T_MAX	/* max value for a size_t */
 
 #define	OFF_MAX		__OFF_MAX	/* max value for an off_t */
 #define	OFF_MIN		__OFF_MIN	/* min value for an off_t */
 #endif
 
-#if __BSD_VISIBLE
+#if 1
 #define	GID_MAX		UINT_MAX	/* max value for a gid_t */
 #define	UID_MAX		UINT_MAX	/* max value for a uid_t */
 
--- a/sys/net80211/ieee80211_ioctl.h
+++ b/sys/net80211/ieee80211_ioctl.h
@@ -575,7 +575,7 @@
 	uint16_t	sv_vlan;
 };
 
-#ifdef __FreeBSD__
+#if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 /*
  * FreeBSD-style ioctls.
  */
@@ -856,6 +856,6 @@
 #define	IEEE80211_CLONE_WDSLEGACY	0x0004	/* legacy WDS processing */
 #define	IEEE80211_CLONE_MACADDR		0x0008	/* use specified mac addr */
 #define	IEEE80211_CLONE_TDMA		0x0010	/* operate in TDMA mode */
-#endif /* __FreeBSD__ */
+#endif /* __FreeBSD__ || __FreeBSD_kernel__ */
 
 #endif /* _NET80211_IEEE80211_IOCTL_H_ */
--- a/sys/net80211/ieee80211_var.h
+++ b/sys/net80211/ieee80211_var.h
@@ -34,7 +34,7 @@
 /* NB: portability glue must go first */
 #if defined(__NetBSD__)
 #include <net80211/ieee80211_netbsd.h>
-#elif defined(__FreeBSD__)
+#elif defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #include <net80211/ieee80211_freebsd.h>
 #elif defined(__linux__)
 #include <net80211/ieee80211_linux.h>
