--- a/sys/sys/queue.h
+++ b/sys/sys/queue.h
@@ -257,8 +257,10 @@
 	struct type **stqh_last;/* addr of last next element */		\
 }
 
+#if defined(__FreeBSD__) || defined(_KERNEL)
 #define	STAILQ_HEAD_INITIALIZER(head)					\
 	{ NULL, &(head).stqh_first }
+#endif
 
 #define	STAILQ_ENTRY(type)						\
 struct {								\
@@ -533,8 +535,10 @@
 }
 #endif
 
+#if defined(__FreeBSD__) || defined(_KERNEL)
 #define	TAILQ_HEAD_INITIALIZER(head)					\
 	{ NULL, &(head).tqh_first, TRACEBUF_INITIALIZER }
+#endif
 
 #if !defined(__GLIBC__) || defined(_KERNEL)
 #define	TAILQ_ENTRY(type)						\
