--- a/sys/sys/queue.h
+++ b/sys/sys/queue.h
@@ -284,8 +284,10 @@
 	class type **stqh_last;	/* addr of last next element */		\
 }
 
+#if defined(__FreeBSD__) || defined(_KERNEL)
 #define	STAILQ_HEAD_INITIALIZER(head)					\
 	{ NULL, &(head).stqh_first }
+#endif
 
 #define	STAILQ_ENTRY(type)						\
 struct {								\
@@ -587,8 +589,10 @@
 }
 #endif
 
+#if defined(__FreeBSD__) || defined(_KERNEL)
 #define	TAILQ_HEAD_INITIALIZER(head)					\
 	{ NULL, &(head).tqh_first, TRACEBUF_INITIALIZER }
+#endif
 
 #if !defined(__GLIBC__) || defined(_KERNEL)
 #define	TAILQ_ENTRY(type)						\
