--- a/sys/sys/socket.h
+++ b/sys/sys/socket.h
@@ -38,6 +38,8 @@
 #include <sys/_iovec.h>
 #include <machine/_align.h>
 
+#if 0
+
 /*
  * Definitions related to sockets: types, address families, options.
  */
@@ -347,6 +349,10 @@
 
 #define	PF_MAX		AF_MAX
 
+#endif /* __BSD_VISIBLE */
+#endif /* 0 */
+
+#if __BSD_VISIBLE
 /*
  * Definitions for network related sysctl, CTL_NET.
  *
@@ -421,6 +427,8 @@
 }
 #endif /* __BSD_VISIBLE */
 
+#if 0
+
 /*
  * Maximum queue length specifiable by listen.
  */
@@ -681,5 +689,6 @@
 
 #endif
 
+#endif
 
 #endif /* !_SYS_SOCKET_H_ */
--- a/sys/sys/_iovec.h
+++ b/sys/sys/_iovec.h
@@ -40,9 +40,13 @@
 #define	_SIZE_T_DECLARED
 #endif
 
+#if 1
+#include <sys/uio.h>
+#else
 struct iovec {
 	void	*iov_base;	/* Base address. */
 	size_t	 iov_len;	/* Length. */
 };
+#endif
 
 #endif /* !_SYS__IOVEC_H_ */
